File: Data Structures & Algorithms/target-sum/submission-5.java
class Solution {

    Map<String, Integer> memo;   

    public int findTargetSumWays(int[] nums, int target) {
        memo = new HashMap<>();
        return solve(nums, 0, target, 0);
    }

    private int solve(int[] nums, int idx, int t, int sum){
        if(idx == nums.length){
            return sum == t? 1:0;
        }

        String key = idx + "-" + sum;
        if(memo.containsKey(key)){
            return memo.get(key);
        }

        int res = solve(nums, idx+1, t, sum - nums[idx]) 
                + solve(nums, idx+1, t, sum + nums[idx]);

        memo.put(key, res);

        return res;
    }
}

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-5.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        total = sum(nums)

        if total % 2 == 1:
            return False

        # target = total/2

        def dfs(i, target):
            if target == 0:
                return True

            if i >= len(nums):
                return False

            return dfs(i+1, target - nums[i]) or dfs(i+1, target)

        return dfs(0, total/2)

